### PR TITLE
httpbp: add middleware for specifying supported HTTP methods

### PR DESCRIPTION
**Pull request by pacejackson**

### CODE REVIEW

File: httpbp/middlewares.go [hunk 1]
```diff
@@ -158,3 +160,31 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
 		}
 	}
 }
+
+// SupportedMethods returns a middleware that checks if the request is made
+// using one of the given HTTP methods.
+//
+// Returns a raw, plain text 405 error response if the method is not supported.
+func SupportedMethods(method string, additional ...string) Middleware {
```
Comment (fishy) [inline httpbp/middlewares.go:168]: 🔕 Is the separation of `method` and `additional` for forcing at least one arg passed in?

File: httpbp/middlewares.go [hunk 2]
```diff
@@ -158,3 +160,31 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
...
+	supported := make(map[string]bool, len(additional)+1)
+	supported[strings.ToUpper(method)] = true
+	for _, m := range additional {
```
Comment (fishy) [inline httpbp/middlewares.go:171]: 🔕 Should we implicitly add HEAD as long as GET is allowed? Go's stdlib http package will handle HEAD requests automatically (drop the body and just return the headers you returned in your handler implementation), and developers are likely to forget add HEAD to the allowed list, and I think in most cases we don't really want to disallow HEAD requests.

File: httpbp/middlewares.go [hunk 3]
```diff
@@ -158,3 +160,31 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
...
+		supported[strings.ToUpper(m)] = true
+	}
+	return func(name string, next HandlerFunc) HandlerFunc {
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+			method := r.Method
+			if method == "" {
+				method = http.MethodGet
+			}
+			if ok := supported[method]; !ok {
```
Comment (fishy) [inline httpbp/middlewares.go:180]: you don't even need `ok` here, just `if supported[method] {` is sufficient :)

File: httpbp/middlewares.go (hunk 1 above)
Comment (pacejackson) [reply httpbp/middlewares.go:168]: Yes

File: httpbp/middlewares.go [hunk 4]
```diff
@@ -158,3 +160,31 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
...
+func SupportedMethods(method string, additional ...string) Middleware {
+	supported := make(map[string]bool, len(additional)+1)
+	supported[strings.ToUpper(method)] = true
+	for _, m := range additional {
+		supported[strings.ToUpper(m)] = true
+	}
+	return func(name string, next HandlerFunc) HandlerFunc {
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+			method := r.Method
+			if method == "" {
```
Comment (fishy) [inline httpbp/middlewares.go:177]: 🔕 In what scenario would this happen?

File: httpbp/middlewares.go (hunk 2 above)
Comment (pacejackson) [reply httpbp/middlewares.go:172]: That makes sense

File: httpbp/middlewares.go [hunk 5]
```diff
@@ -158,3 +160,40 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
...
+	supported := make(map[string]bool, len(additional)+1)
+	supported[strings.ToUpper(method)] = true
+	for _, m := range additional {
+		supported[strings.ToUpper(m)] = true
+	}
+
+	hasGet := false
+	for m := range supported {
+		hasGet = hasGet || strings.Compare(m, http.MethodGet) == 0
+	}
```
Comment (fishy) [inline httpbp/middlewares.go:179]: you don't need this for loop. you already built the map, so you can just do a map lookup :)

but implicitly adding HEAD means that if some service really want to allow GET and disallow HEAD, there's no way they can achieve that with this middleware. I don't really know if that's a legit case in real life, though.

File: httpbp/middlewares.go (hunk 4 above)
Comment (pacejackson) [reply httpbp/middlewares.go:186]: I realized that I just misinterpreted the documentation for this:

```go
    // Method specifies the HTTP method (GET, POST, PUT, etc.).
    // For client requests, an empty string means GET.
    //
    // Go's HTTP client does not support sending a request with
    // the CONNECT method. See the documentation on Transport for
    // details.
    Method string
```

I don't think this should happen on the server side so I can probably remove this.

File: httpbp/middlewares.go (hunk 5 above)
Comment (pacejackson) [reply httpbp/middlewares.go]: 🤦 

As far as not allowing HEAD... I'm not sure there's _really_ a use case for that and if there is, that endpoint can just not use this middleware and make their own.

**Comment by pacejackson** [comment]
You're also supposed to return an `Allow` header with a list of the supported HTTP methods so I added that as well.

File: httpbp/middlewares.go [hunk 6]
```diff
@@ -158,3 +165,43 @@ func InjectEdgeRequestContext(truster HeaderTrustHandler, impl *edgecontext.Impl
...
+	for m := range supported {
+		allowed[i] = m
+		i++
+	}
+	sort.Strings(allowed)
+	allowedHeader := strings.Join(allowed, ",")
+
+	return func(name string, next HandlerFunc) HandlerFunc {
+		return func(ctx context.Context, w http.ResponseWriter, r *http.Request) error {
+			if ok := supported[r.Method]; !ok {
```
Comment (fishy) [inline httpbp/middlewares.go:196]: nit: Here it can be just `if !supported[r.Method] {`

Comment (pacejackson) [reply httpbp/middlewares.go]: ahh, meant to fix that 😬

**Pull request merged**